src/votekit/ballot_generator.py

```diff
@@ -509,84 +515,110 @@ class AlternatingCrossover(BallotGenerator):
+        
+        voter_props = [self.cohesion_parameters[b]*self.bloc_voter_prop[b] if t == "bloc" \
+                       else (1-self.cohesion_parameters[b])*self.bloc_voter_prop[b]\
+                       for b,t in voter_types]
+
+        ballots_per_type = dict(zip(voter_types, apportion.compute("huntington", voter_props, 
+                                                              number_of_ballots)))
+
+
+
         for bloc in self.bloc_voter_prop.keys():
 
-            num_ballots = self.round_num(number_of_ballots * self.bloc_voter_prop[bloc])
-            crossover_dict = self.bloc_crossover_rate[bloc]
+            num_bloc_ballots = ballots_per_type[(bloc, "bloc")]
+            num_cross_ballots = ballots_per_type[(bloc, "cross")]
+
             pref_interval_dict = self.pref_interval_by_bloc[bloc]
 
-            # generates crossover ballots from each bloc (allowing for more than two blocs)
-            for opposing_slate in crossover_dict.keys():
-                crossover_rate = crossover_dict[opposing_slate]
-                num_crossover_ballots = self.round_num(crossover_rate * num_ballots)
-
-                opposing_cands = self.slate_to_candidates[opposing_slate]
-                bloc_cands = self.slate_to_candidates[bloc]
-
-                for _ in range(num_crossover_ballots):
-                    pref_for_opposing = [
-                        pref_interval_dict[cand] for cand in opposing_cands
-                    ]
-                    # convert to probability distribution
-                    pref_for_opposing = [
-                        p / sum(pref_for_opposing) for p in pref_for_opposing
-                    ]
-
-                    pref_for_bloc = [pref_interval_dict[cand] for cand in bloc_cands]
-                    # convert to probability distribution
-                    pref_for_bloc = [p / sum(pref_for_bloc) for p in pref_for_bloc]
-
-                    bloc_cands = list(
-                        np.random.choice(
-                            bloc_cands,
-                            p=pref_for_bloc,
-                            size=len(bloc_cands),
-                            replace=False,
-                        )
+            opposing_slate = list(set(self.bloc_voter_prop.keys()).difference([bloc]))[0]
+            opposing_cands = self.slate_to_candidates[opposing_slate]
+            bloc_cands = self.slate_to_candidates[bloc]
+
+            pref_for_opposing = [
+                    pref_interval_dict[cand] for cand in opposing_cands
+                ]
+            # convert to probability distribution
+            pref_for_opposing = [
+                p / sum(pref_for_opposing) for p in pref_for_opposing
+            ]
+
+            pref_for_bloc = [pref_interval_dict[cand] for cand in bloc_cands]
+            # convert to probability distribution
+            pref_for_bloc = [p / sum(pref_for_bloc) for p in pref_for_bloc]
+
+            for _ in range(num_cross_ballots):
```

Comment:
Why are there two for loops (one for `range(num_cross_ballots)` and another  for `range(num_cross_ballots)`), it looks like the `ranking` is all that is changing between them, but correct me if wrong!

Could we do `for i in range (num_cross_ballots + num_bloc_ballots)` and then use a conditional based on the number of ballots generated to switch whether the ranking is a crossover ballot or an opposing?